### PR TITLE
Added Validator::atLeast and atMost

### PR DESCRIPTION
**Pull request by lorenzo**
Improves on #8716

### CODE REVIEW

File: src/Validation/Validator.php [hunk 1]
```diff
@@ -1320,6 +1320,56 @@ public function multipleOptions($field, array $options = [], $message = null, $w
...
+     *
+     * @param string $field The field you want to apply the rule to.
+     * @param int $count The number of elements the array should at least have
+     * @param string|null $message The error message when the rule fails.
+     * @param string|callable|null $when Either 'create' or 'update' or a callable that returns
+     *   true when the validation rule should be applied.
+     * @see \Cake\Validation\Validation::numElements()
+     * @return $this
+     */
+    public function hasAtLeast($field, $count, $message = null, $when = null)
```
Comment (ravage84) [inline src/Validation/Validator.php]: The code of ``hasAtLeast()`` and ``hasAtMost()`` almost the same.
Besides the name of the method, only two things differ:
1. ``'hasAtLeast'`` vs. ``'hasAtMost'``
2. ``'>=' ``vs. ``'<='``

I wonder if that could be de-duplicated somehow.

File: src/Validation/Validator.php [hunk 2]
```diff
@@ -1320,6 +1320,56 @@ public function multipleOptions($field, array $options = [], $message = null, $w
...
+     * @param string|null $message The error message when the rule fails.
+     * @param string|callable|null $when Either 'create' or 'update' or a callable that returns
+     *   true when the validation rule should be applied.
+     * @see \Cake\Validation\Validation::numElements()
+     * @return $this
+     */
+    public function hasAtMost($field, $count, $message = null, $when = null)
+    {
+        $extra = array_filter(['on' => $when, 'message' => $message]);
+        return $this->add($field, 'hasAtMost', $extra + [
```
Comment (ravage84) [inline src/Validation/Validator.php]: Haven't worked with CakePHP 3's Validation yet, but this whole method does not look very readable to me, at a first glance.

Comment (markstory) [inline src/Validation/Validator.php]: Many of this class' wrapper functions are not super readable, but let application code be simpler. This method in use would look like:

```php
$validator->hasAtMost('user_id', 3)
```

**Comment by lorenzo** [comment]
Ok if I merge this now?

**Pull request merged**